File: app/test/post_media_test.dart
import 'dart:convert';

import 'package:checkin/api/models.dart';
import 'package:flutter_test/flutter_test.dart';

/// The client half of the media wire-format pair. Its twin is
/// `TestPostMediaWireFormat` in server/internal/db/postmedia_test.go, which pins the same
/// literal from the marshalling end: a field renamed or retyped on either side fails the
/// pair here rather than shipping a feed the app renders as broken images.
void main() {
  test('parses the media array the server pins in TestPostMediaWireFormat', () {
    const wire = '[{"id":7,"mime":"image/jpeg","width":1600,"height":1200,"durationMs":0,'
        '"hasPoster":false},{"id":8,"mime":"video/mp4","width":1080,"height":1920,'
        '"durationMs":9500,"hasPoster":true}]';

    final media = (jsonDecode(wire) as List)
        .map((e) => PostMedia.fromJson(e as Map<String, dynamic>))
        .toList();

    expect(media.length, 2);

    expect(media[0].id, 7);
    expect(media[0].mime, 'image/jpeg');
    expect(media[0].width, 1600);
    expect(media[0].height, 1200);
    expect(media[0].durationMs, 0);
    expect(media[0].hasPoster, isFalse);
    expect(media[0].isImage, isTrue);
    expect(media[0].isVideo, isFalse);

    expect(media[1].id, 8);
    expect(media[1].mime, 'video/mp4');
    expect(media[1].width, 1080);
    expect(media[1].height, 1920);
    expect(media[1].durationMs, 9500);
    expect(media[1].hasPoster, isTrue);
    expect(media[1].isVideo, isTrue);
    expect(media[1].isImage, isFalse);
  });

  test('a post carries its attachments typed and in order', () {
    final post = Post.fromJson({
      'id': 1,
      'authorId': 2,
      'kind': 'video',
      'createdAt': '2026-08-14T09:00:00Z',
      'mediaIds': [7, 8],
      'media': [
        {'id': 7, 'mime': 'image/gif', 'width': 400, 'height': 400},
        {'id': 8, 'mime': 'video/mp4', 'width': 1080, 'height': 1920, 'durationMs': 9500},
      ],
    });

    expect([for (final m in post.media) m.id], [7, 8]);
    expect(post.media[0].isGif, isTrue);
    expect(post.media[1].isVideo, isTrue);
    // mediaIds keeps its own meaning: published clients read only that list.
    expect(post.mediaIds, [7, 8]);
    expect(post.images, [7, 8]);
  });

  test('an old server sending only mediaIds still yields image attachments', () {
    final post = Post.fromJson({
      'id': 1,
      'authorId': 2,
      'kind': 'image',
      'createdAt': '2026-08-14T09:00:00Z',
      'mediaIds': [3, 4],
    });

    expect([for (final m in post.media) m.id], [3, 4]);
    expect(post.media.every((m) => m.isImage), isTrue);
    expect(post.media.every((m) => m.isVideo), isFalse);
  });

  test('an old server sending only the legacy cover yields one image attachment', () {
    final post = Post.fromJson({
      'id': 1,
      'authorId': 2,
      'kind': 'image',
      'createdAt': '2026-08-14T09:00:00Z',
      'mediaId': 42,
    });

    expect(post.media.length, 1);
    expect(post.media.single.id, 42);
    expect(post.media.single.isImage, isTrue);
  });

  test('a text post has no attachments at all', () {
    final post = Post.fromJson({
      'id': 1,
      'authorId': 2,
      'kind': 'text',
      'body': 'hello',
      'createdAt': '2026-08-14T09:00:00Z',
    });

    expect(post.media, isEmpty);
    expect(post.imageMedia, isEmpty);
  });

  test('only images are offered to the gallery', () {
    final post = Post.fromJson({
      'id': 1,
      'authorId': 2,
      'kind': 'video',
      'createdAt': '2026-08-14T09:00:00Z',
      'media': [
        {'id': 8, 'mime': 'video/mp4', 'durationMs': 9500},
        {'id': 9, 'mime': 'image/jpeg'},
      ],
    });

    expect([for (final m in post.imageMedia) m.id], [9]);
  });

  test('a clip reports its length as m:ss, rounded up so it is never 0:00', () {
    String label(int ms) => PostMedia(id: 1, mime: 'video/mp4', durationMs: ms).durationLabel;
    expect(label(9500), '0:10');
    expect(label(9000), '0:09');
    expect(label(200), '0:01');
    expect(label(62000), '1:02');
    // A still is not a timed medium, so it has nothing to show.
    expect(const PostMedia(id: 1, mime: 'image/jpeg').durationLabel, '');
  });

  test('stored dimensions give a ratio before anything has decoded', () {
    expect(const PostMedia(id: 1, mime: 'video/mp4', width: 1080, height: 1920).aspectRatio,
        closeTo(1080 / 1920, 0.0001));
    expect(const PostMedia(id: 1, mime: 'image/jpeg').aspectRatio, isNull);
  });

  test('ServerInfo reads the media types a server accepts', () {
    final modern = ServerInfo.fromJson({
      'name': 'Alpha',
      'initialized': true,
      'mediaTypes': ['image', 'gif', 'video'],
    });
    expect(modern.mediaTypes, ['image', 'gif', 'video']);

    // A server predating typed media only ever accepted stills, so the absent key must not
    // be read as "anything goes" - it is what stops a clip being offered to a server that
    // would reject it.
    final old = ServerInfo.fromJson({'name': 'Beta', 'initialized': true});
    expect(old.mediaTypes, ['image']);
  });
}

File: server/internal/db/postmedia_test.go
package db

import (
	"encoding/json"
	"strings"
	"testing"
)

// The post payload's media array, pinned from this end. Its twin lands with the client
// pass (app/test, feeding this same literal to the Dart parser), so a shape change on one
// side fails the pair rather than shipping a feed the app renders as broken images.
func TestPostMediaWireFormat(t *testing.T) {
	const want = `[{"id":7,"mime":"image/jpeg","width":1600,"height":1200,"durationMs":0,"hasPoster":false},` +
		`{"id":8,"mime":"video/mp4","width":1080,"height":1920,"durationMs":9500,"hasPoster":true}]`

	got, err := json.Marshal([]PostMedia{
		{ID: 7, Mime: "image/jpeg", Width: 1600, Height: 1200},
		{ID: 8, Mime: "video/mp4", Width: 1080, Height: 1920, DurationMs: 9500, HasPoster: true},
	})
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	if string(got) != want {
		t.Errorf("media json =\n%s\nwant\n%s", got, want)
	}
}

// mediaIds is what every published client reads, and it has to keep coming out of the same
// query as the typed array or a client on one and a client on the other see different posts.
func TestApplyMediaDerivesTheLegacyIDList(t *testing.T) {
	var p Post
	p.applyMedia([]byte(`[{"id":3,"mime":"video/mp4"},{"id":1,"mime":"image/jpeg"}]`))

	if len(p.Media) != 2 {
		t.Fatalf("Media has %d entries, want 2", len(p.Media))
	}
	if len(p.MediaIDs) != 2 || p.MediaIDs[0] != 3 || p.MediaIDs[1] != 1 {
		t.Errorf("MediaIDs = %v, want [3 1] in the array's order", p.MediaIDs)
	}
}

// A text post must not grow an empty array in its payload: omitempty only drops a nil one,
// and older clients treat a present-but-empty list as "this post has media".
func TestApplyMediaLeavesTextPostsBare(t *testing.T) {
	var p Post
	p.applyMedia([]byte(`[]`))

	if p.MediaIDs != nil || p.Media != nil {
		t.Fatalf("MediaIDs = %v, Media = %v, want both nil", p.MediaIDs, p.Media)
	}
	out, err := json.Marshal(p)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	if strings.Contains(string(out), "media") {
		t.Errorf("text post payload mentions media: %s", out)
	}
}

// The stored kind describes what is attached, whatever the client called the post.
func TestKindFor(t *testing.T) {
	image := PostMedia{ID: 1, Mime: "image/jpeg"}
	gif := PostMedia{ID: 2, Mime: "image/gif"}
	video := PostMedia{ID: 3, Mime: "video/mp4"}

	tests := []struct {
		name  string
		media []PostMedia
		want  string
	}{
		{"nothing attached is a text post", nil, "text"},
		{"a photo is an image post", []PostMedia{image}, "image"},
		{"a gif is still an image post", []PostMedia{gif}, "image"},
		{"a clip is a video post", []PostMedia{video}, "video"},
		{"one clip among photos makes it a video post", []PostMedia{image, video, gif}, "video"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := kindFor(tt.media); got != tt.want {
				t.Errorf("kindFor() = %q, want %q", got, tt.want)
			}
		})
	}
}

// The legacy cover must never point at a clip: published clients render posts.media_id
// as a picture, so a video there paints a broken-image icon where caption-only is the
// intended degradation.
func TestCoverForSkipsClips(t *testing.T) {
	clipFirst := []PostMedia{
		{ID: 8, Mime: "video/mp4"},
		{ID: 7, Mime: "image/jpeg"},
	}
	if got := coverFor(clipFirst); got == nil || *got != 7 {
		t.Errorf("coverFor(clip first) = %v, want the first image id 7", got)
	}
	clipOnly := []PostMedia{{ID: 8, Mime: "video/mp4"}}
	if got := coverFor(clipOnly); got != nil {
		t.Errorf("coverFor(clip only) = %v, want nil", got)
	}
	if got := coverFor(nil); got != nil {
		t.Errorf("coverFor(nil) = %v, want nil", got)
	}
}
